File: app/src/main/java/com/bigscreen/mangindo/newrelease/NewReleaseAdapter.java
package com.bigscreen.mangindo.newrelease;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bigscreen.mangindo.R;
import com.bigscreen.mangindo.common.Constant;
import com.bigscreen.mangindo.stored.StoredDataService;
import com.bigscreen.mangindo.listener.OnListItemClickListener;
import com.bigscreen.mangindo.listener.OnLoadDataListener;
import com.bigscreen.mangindo.network.model.Manga;
import com.bigscreen.mangindo.network.model.response.NewReleaseResponse;
import com.bigscreen.mangindo.network.service.MangaApiService;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class NewReleaseAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements
        NewReleaseLoader.OnLoadNewReleaseListener, NewReleaseViewHolder.OnMangaClickListener {

    public static final int SORT_BY_DATE = 1;
    public static final int SORT_BY_TITLE = 2;

    private Context context;

    private List<Manga> mangaList, backupMangaList;
    private NewReleaseLoader newReleaseLoader;
    private OnLoadDataListener loadDataListener;
    private OnListItemClickListener listItemClickListener;
    private StoredDataService storedDataService;

    private int sortBy = SORT_BY_DATE;
    private String searchKeyword = "";

    public NewReleaseAdapter(Context context, OnLoadDataListener loadDataListener,
                             StoredDataService storedDataService, MangaApiService apiService) {
        this.context = context;
        this.loadDataListener = loadDataListener;
        this.storedDataService = storedDataService;
        newReleaseLoader = new NewReleaseLoader(apiService, this);
    }

    public void loadManga() {
        newReleaseLoader.loadNewReleaseManga();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_new_release, parent, false);
        return new NewReleaseViewHolder(view, context, this);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (searchKeyword.isEmpty())
            ((NewReleaseViewHolder) holder).bindData(getItem(position));
        else
            ((NewReleaseViewHolder) holder).bindDataSearch(getItem(position), searchKeyword);
    }

    @Override
    public int getItemCount() {
        return mangaList == null ? 0 : mangaList.size();
    }

    public Manga getItem(int position) {
        return getItemCount() > 0 ? mangaList.get(position) : null;
    }

    private void setMangaList(List<Manga> mangaList) {
        this.mangaList = mangaList;
        backupMangaList = mangaList;
        notifyDataSetChanged();
    }

    @Override
    public void onPrepareLoadData() {
        Log.d(Constant.LOG_TAG, "load data...");
        loadDataListener.onPrepare();
    }

    @Override
    public void onSuccessLoadData(NewReleaseResponse newReleaseResponse) {
        Log.d(Constant.LOG_TAG, "success load data\n" + newReleaseResponse.toString());
        loadDataListener.onSuccess();
        setMangaList(newReleaseResponse.getComics());
        storedDataService.saveNewReleasedComic(newReleaseResponse);
    }

    @Override
    public void onFailedLoadData(String message) {
        Log.e(Constant.LOG_TAG, "failed load data, " + message);
        loadDataFromPreference(message);
    }

    private void loadDataFromPreference(final String networkErrorMessage) {
        storedDataService.pullStoredNewReleasedComic(new StoredDataService.OnGetSavedDataListener<NewReleaseResponse>() {
            @Override
            public void onDataFound(NewReleaseResponse savedData) {
                loadDataListener.onSuccess();
                setMangaList(savedData.getComics());
            }

            @Override
            public void onDataNotFound() {
                loadDataListener.onError(networkErrorMessage);
            }
        });
    }

    public void onParentDestroyed() {
        newReleaseLoader.unSubscribe();
    }

    @Override
    public void onMangaClick(int position) {
        if (listItemClickListener != null)
            listItemClickListener.onListItemClick(position);
    }

    public void setOnListItemClickListener(OnListItemClickListener listItemClickListener) {
        this.listItemClickListener = listItemClickListener;
    }

    public void setSortingOption(int sortBy) {
        if (this.sortBy == sortBy) return;
        this.sortBy = sortBy;
        switch (sortBy) {
            case SORT_BY_DATE:
                sortMangaListByDate();
                break;
            case SORT_BY_TITLE:
                sortMangaListByName();
                break;
            default:
                break;
        }
    }

    private void sortMangaListByDate() {
        Collections.sort(mangaList, new Comparator<Manga>() {
            @Override
            public int compare(Manga manga, Manga mangaComparator) {
                int id = manga.getId();
                int idComparator = mangaComparator.getId();
                return id - idComparator;
            }
        });
        notifyDataSetChanged();
    }

    private void sortMangaListByName() {
        Collections.sort(mangaList, new Comparator<Manga>() {
            @Override
            public int compare(Manga manga, Manga mangaComparator) {
                String title = manga.getTitle().toLowerCase();
                String titleComparator = mangaComparator.getTitle().toLowerCase();
                return title.compareTo(titleComparator);
            }
        });
        notifyDataSetChanged();
    }

    public void filterList(String keyword) {
        searchKeyword = keyword;
        mangaList = getBackupMangaList();
        if (!keyword.isEmpty()) {
            List<Manga> filteredList = new ArrayList<>();
            for (Manga manga : mangaList) {
                if (isStartWith(manga.getTitle(), keyword) || isStartWith(manga.getHiddenComic(), keyword)) {
                    filteredList.add(manga);
                }
            }
            mangaList = filteredList;
        }
        notifyDataSetChanged();
    }

    private boolean isStartWith(String firstArg, String secondArg) {
        return firstArg.toLowerCase().startsWith(secondArg.toLowerCase());
    }

    private List<Manga> getBackupMangaList() {
        List<Manga> tempMangaList = backupMangaList;
        switch (sortBy) {
            case SORT_BY_DATE:
                Collections.sort(tempMangaList, new Comparator<Manga>() {
                    @Override
                    public int compare(Manga manga, Manga mangaComparator) {
                        int id = manga.getId();
                        int idComparator = mangaComparator.getId();
                        return id - idComparator;
                    }
                });
                break;
            case SORT_BY_TITLE:
                Collections.sort(tempMangaList, new Comparator<Manga>() {
                    @Override
                    public int compare(Manga manga, Manga mangaComparator) {
                        String title = manga.getTitle().toLowerCase();
                        String titleComparator = mangaComparator.getTitle().toLowerCase();
                        return title.compareTo(titleComparator);
                    }
                });
                break;
            default:
                break;
        }
        return tempMangaList;
    }

}

File: app/src/main/java/com/bigscreen/mangindo/newrelease/NewReleaseAdapter.kt
package com.bigscreen.mangindo.newrelease

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bigscreen.mangindo.R
import com.bigscreen.mangindo.stored.StoredDataService
import com.bigscreen.mangindo.listener.OnListItemClickListener
import com.bigscreen.mangindo.listener.OnLoadDataListener
import com.bigscreen.mangindo.network.model.Manga
import com.bigscreen.mangindo.network.model.response.NewReleaseResponse
import com.bigscreen.mangindo.network.service.MangaApiService
import java.util.Collections

class NewReleaseAdapter(
        private val context: Context,
        private val loadDataListener: OnLoadDataListener,
        private val storedDataService: StoredDataService,
        apiService: MangaApiService
) : RecyclerView.Adapter<RecyclerView.ViewHolder>(),
        NewReleaseLoader.OnLoadNewReleaseListener,
        NewReleaseViewHolder.OnMangaClickListener {

    companion object {
        const val SORT_BY_DATE = 1
        const val SORT_BY_TITLE = 2
    }

    private val newReleaseLoader = NewReleaseLoader(apiService, this)
    private var backupMangaList = emptyList<Manga>()
    private var mangaList = emptyList<Manga>()
        set(value) {
            field = value
            backupMangaList = value
            notifyDataSetChanged()
        }

    var listItemClickListener: OnListItemClickListener? = null

    private var sortBy = SORT_BY_DATE
    private var searchKeyword = ""

    fun loadManga() {
        newReleaseLoader.loadNewReleaseManga()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.item_new_release, parent, false)
        return NewReleaseViewHolder(view, context, this)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (searchKeyword.isEmpty())
            (holder as NewReleaseViewHolder).bindData(getItem(position))
        else
            (holder as NewReleaseViewHolder).bindDataSearch(getItem(position), searchKeyword)
    }

    override fun getItemCount(): Int = mangaList.size

    override fun onPrepareLoadData() {
        loadDataListener.onPrepare()
    }

    override fun onSuccessLoadData(newReleaseResponse: NewReleaseResponse) {
        loadDataListener.onSuccess()
        mangaList = newReleaseResponse.comics
        storedDataService.saveNewReleasedComic(newReleaseResponse)
    }

    override fun onFailedLoadData(message: String) {
        loadDataFromPreference(message)
    }

    override fun onMangaClick(position: Int) {
        listItemClickListener?.onListItemClick(position)
    }

    fun getItem(position: Int): Manga = mangaList[position]

    fun onParentDestroyed() {
        newReleaseLoader.unSubscribe()
    }

    fun filterList(keyword: String) {
        searchKeyword = keyword
        mangaList = getBackupMangaList()
        if (keyword.isNotEmpty()) {
            mangaList = mangaList.filter { isStartWith(it.title, keyword) || isStartWith(it.hiddenComic, keyword) }
        }
        notifyDataSetChanged()
    }

    private fun loadDataFromPreference(networkErrorMessage: String) {
        storedDataService.pullStoredNewReleasedComic(object : StoredDataService.OnGetSavedDataListener<NewReleaseResponse> {
            override fun onDataFound(savedData: NewReleaseResponse) {
                loadDataListener.onSuccess()
                mangaList = savedData.comics
            }

            override fun onDataNotFound() {
                loadDataListener.onError(networkErrorMessage)
            }
        })
    }

    fun setSortingOption(sortBy: Int) {
        if (this.sortBy == sortBy) return
        this.sortBy = sortBy
        when (sortBy) {
            SORT_BY_DATE -> sortMangaListByDate()
            SORT_BY_TITLE -> sortMangaListByName()
        }
    }

    private fun sortMangaListByDate() {
        Collections.sort(mangaList) { manga, mangaComparator ->
            val id = manga?.id ?: 0
            val idComparator = mangaComparator?.id ?: 0
            id - idComparator
        }
        notifyDataSetChanged()
    }

    private fun sortMangaListByName() {
        Collections.sort(mangaList) { manga, mangaComparator ->
            val title = manga?.title?.toLowerCase() ?: ""
            val titleComparator = mangaComparator?.title?.toLowerCase() ?: ""
            title.compareTo(titleComparator)
        }
        notifyDataSetChanged()
    }

    private fun isStartWith(firstArg: String, secondArg: String): Boolean {
        return firstArg.toLowerCase().startsWith(secondArg.toLowerCase())
    }

    private fun getBackupMangaList(): List<Manga> {
        return when (sortBy) {
            SORT_BY_DATE -> {
                backupMangaList.sortedWith(Comparator { manga, mangaComparator ->
                    val id = manga?.id ?: 0
                    val idComparator = mangaComparator?.id ?: 0
                    id - idComparator
                })
            }
            SORT_BY_TITLE -> {
                backupMangaList.sortedWith(Comparator { manga, mangaComparator ->
                    val title = manga?.title?.toLowerCase() ?: ""
                    val titleComparator = mangaComparator?.title?.toLowerCase() ?: ""
                    title.compareTo(titleComparator)
                })
            }
            else -> emptyList()
        }
    }

}

